cartography/intel/azure/cosmosdb.py

```diff
@@ -0,0 +1,1090 @@
+        return []
+    except HttpResponseError as e:
+        logger.warning(f"Error while retrieving Cassandra keyspaces list - {e}")
+        return []
+
+    return cassandra_keyspace_list
+
+
+@timeit
+def get_mongodb_databases(credentials: Credentials, subscription_id: str, database_account: Dict) -> List[Dict]:
+    """
+    Return the list of MongoDB Databases in a database account.
+    """
+    try:
+        client = get_client(credentials, subscription_id)
+        mongodb_database_list = list(
+            map(
+                lambda x: x.as_dict(),
+                client.mongo_db_resources.list_mongo_db_databases(
+                    database_account['resourceGroup'],
+                    database_account['name'],
+                ),
+            ),
+        )
+
+    except ClientAuthenticationError as e:
+        logger.warning(f"Client Authentication Error while retrieving MongoDB Databases - {e}")
+        return []
+    except ResourceNotFoundError as e:
+        logger.warning(f"MongoDB Database resource not found error - {e}")
+        return []
+    except HttpResponseError as e:
+        logger.warning(f"Error while retrieving MongoDB Database list - {e}")
+        return []
+
+    return mongodb_database_list
+
+
+@timeit
+def get_table_resources(credentials: Credentials, subscription_id: str, database_account: Dict) -> List[Dict]:
+    """
+    Return the list of Table Resources in a database account.
+    """
+    try:
+        client = get_client(credentials, subscription_id)
+        table_resources_list = list(
+            map(
+                lambda x: x.as_dict(),
+                client.table_resources.list_tables(
+                    database_account['resourceGroup'],
+                    database_account['name'],
+                ),
+            ),
+        )
+
+    except ClientAuthenticationError as e:
+        logger.warning(f"Client Authentication Error while retrieving table resources - {e}")
+        return []
+    except ResourceNotFoundError as e:
+        logger.warning(f"Table resource not found error - {e}")
+        return []
+    except HttpResponseError as e:
+        logger.warning(f"Error while retrieving table resources list - {e}")
+        return []
+
+    return table_resources_list
+
+
+@timeit
+def load_database_account_details(
+        neo4j_session: neo4j.Session, credentials: Credentials, subscription_id: str,
+        details: List[Tuple[Any, Any, Any, Any, Any, Any, Any]], update_tag: int, common_job_parameters: Dict,
+) -> None:
+    """
+    Create dictionaries for SQL Databases, Cassandra Keyspaces, MongoDB Databases and table resources.
+    """
+    sql_databases: List[Dict] = []
+    cassandra_keyspaces: List[Dict] = []
+    mongodb_databases: List[Dict] = []
+    table_resources: List[Dict] = []
+
+    for account_id, name, resourceGroup, sql_database, cassandra_keyspace, mongodb_database, table in details:
```

Comment:
Same comment here about decoupling the data transforms from the load functions.

I also  think it's more maintainable to split these by resource type: one function for sql dbs, one for cassandra, one for mongo, one for tables.

Reply:
@achantavy - I added a transform function for the manipulations happening here. My bad I forgot to do that.

About the splitting up bit, firstly, it would just repeat a lot of code. I have split it up later for the individual resources under SQL databases, Cassandra keyspaces and Mongo. Second, these are the direct resources under Database Accounts  (our primary resource) so I figured I'll keep them in this function  as similar conventions were followed in the AWS codes. Lastly, the `details` gets populated in `get_database_account_details()` (line 436). If I have to split up the code for each resource and I'll have to do some major overhauls which I feel might repeat the code.

Reply:
Fair points on copying prior patterns. Disclaimer: there are lots of prior code in the AWS modules that aren't exactly shining examples of great software :-)

In general, I think it's less bad to repeat code than it is to tightly couple and generalize functions too early ("write code that is easy to delete" and all that).

All that said, we're working to improve things by standardizing on the load steps in the near future: https://docs.google.com/document/d/1IZ12R3oROn11LcYj5XunokyOjJkKu-H2O1TEk065Dsk/edit#

Will not block for now.

Reply:
Alright. I'll def check out this document :) Thanks for letting me know :)